src/entities/subscription-base.ts

```diff
@@ -0,0 +1,561 @@
+import { EventDispatcher, Listener } from '../core/components/event-dispatcher';
+import { SubscriptionOptions } from './interfaces/subscription-capable';
+import { EventHandleCapable } from './interfaces/event-handle-capable';
+import { EventEmitCapable } from './interfaces/event-emit-capable';
+import type { PubNubCore as PubNub } from '../core/pubnub-common';
+import * as Subscription from '../core/types/api/subscription';
+import uuidGenerator from '../core/components/uuid';
+
+/**
+ * Subscription state object.
+ *
+ * State object used across multiple subscription object clones.
+ *
+ * @internal
+ */
+export abstract class SubscriptionBaseState {
+  /**
+   * PubNub instance which will perform subscribe / unsubscribe requests.
+   */
+  client: PubNub<unknown, unknown>;
+
+  /**
+   * Whether a subscribable object subscribed or not.
+   */
+  _isSubscribed: boolean = false;
+
+  /**
+   * User-provided a subscription input object.
+   */
+  readonly subscriptionInput: Subscription.SubscriptionInput;
+
+  /**
+   * High-precision timetoken of the moment when subscription was created for entity.
+   *
+   * **Note:** This value can be set only if recently there were active subscription loop.
+   */
+  readonly referenceTimetoken?: string;
+
+  /**
+   * Subscription time cursor.
+   */
+  cursor?: Subscription.SubscriptionCursor;
+
+  /**
+   * SubscriptionCapable object configuration.
+   */
+  options?: SubscriptionOptions;
+
+  /**
+   * The list of references to all {@link SubscriptionBase} clones created for this reference.
+   */
+  clones: Record<string, SubscriptionBase> = {};
```

Comment:
Trying to understand it. What are the exact differences between `cloneEmpty` and `clone`.
duplicating hooked callbacks are the only difference between them or something else also?   Does it has to to do with any SubscriptionBade link? when dealing with any clone method types?